woocommerce/woocommerce-gutenberg-products-block
# Check if cart needs payment before showing payment methods

Fixes #2055.

### Screenshots

![imatge](https://user-images.githubusercontent.com/3616980/78787597-dd27dd80-79aa-11ea-8955-2802a7c1095a.png)

### How to test the changes in this Pull Request:

1. Create a product with price 0.
2. When selecting a shipping method with price 0, verify payment methods are not rendered in the _Checkout_ block.
3. When selecting a shipping method with price >0, verify payment methods appear again.
4. Try completing the purchase and verify it works.

### Concerns

Some of the concerns expressed in #2055:

> re-rendering payment methods every time a shipping rates change (selected or generated from address change) could be expensive (emphasis on could).

That will only happen when total price changes from `0` to `>0`. Even though this can certainly occur, I don't think it's something that will happen repeatedly for normal sessions: shipping rates are debounced when shipping address is introduced and users will probably interact with the shipping rates selector once.

> Views could be janky because the containers for the payment methods would appear/disappear depending on the shipping rate selections. This would be even more pronounced with mobile.

I could reproduce this with Firefox and Gnome Web (which uses WebKit, like Safari). Chrome seems to anchor the view to the clicked form element so there are no visual jumps.

I was playing around with [`overflow-anchor`](https://developer.mozilla.org/en-US/docs/Web/CSS/overflow-anchor) with no luck. So my vote would be to leave it as-is: behavior seems to depend on the user agent and as exposed above, I don't think payment methods will appear/disappear repeatedly in a normal customer flow.
